Source/SwiftLintFramework/Reporters/SummaryReporter.swift

```diff
@@ -0,0 +1,113 @@
+import Foundation
+import SwiftyTextTable
+
+/// Reports a summary table of all violations
+public struct SummaryReporter: Reporter {
+    // MARK: - Reporter Conformance
+
+    public static let identifier = "summary"
+    public static let isRealtime = false
+
+    public var description: String {
+        return "Reports a summary table of all violations."
+    }
+
+    public static func generateReport(_ violations: [StyleViolation]) -> String {
+        TextTable(violations: violations).renderWithExtraSeparator()
+    }
+}
+
+// MARK: - SwiftyTextTable
+
+private extension TextTable {
+    init(violations: [StyleViolation]) {
+        let numberOfViolationHeader = "number of violations"
+        let numberOfFileHeader = "number of files"
+        let columns = [
+            TextTableColumn(header: "rule identifier"),
+            TextTableColumn(header: "opt-in"),
+            TextTableColumn(header: "correctable"),
+            TextTableColumn(header: "custom"),
+            TextTableColumn(header: numberOfViolationHeader),
+            TextTableColumn(header: numberOfFileHeader)
+        ]
+        self.init(columns: columns)
+
+        let ruleIdentifiersToViolationsMap = violations.group { $0.ruleIdentifier }
+        let sortedRuleIdentifiers = ruleIdentifiersToViolationsMap.keys.sorted {
+            let count1 = ruleIdentifiersToViolationsMap[$0]?.count ?? 0
+            let count2 = ruleIdentifiersToViolationsMap[$1]?.count ?? 0
+            if count1 > count2 {
+                return true
+            } else if count1 == count2 {
+                return $0 < $1
+            }
+            return false
+        }
+
+        var totalNumberOfViolations = 0
+
+        for ruleIdentifier in sortedRuleIdentifiers {
+            guard let ruleIdentifier = ruleIdentifiersToViolationsMap[ruleIdentifier]?.first?.ruleIdentifier else {
+                continue
+            }
+
+            var rule: Rule?
+            if let ruleType = primaryRuleList.list[ruleIdentifier] {
```

Comment:
You can just `let rule = primaryRuleList.list[ruleIdentifier]` and then check `rule is OptInRule.Type`.